p2p/net/conn/listen.go

```diff
@@ -144,9 +145,9 @@ func Listen(ctx context.Context, addr ma.Multiaddr, local peer.ID, sk ic.PrivKey
 		Listener: ml,
 		local:    local,
 		privk:    sk,
-		cg:       ctxgroup.WithContext(ctx),
+		proc:     goprocessctx.WithContext(ctx),
 	}
-	l.cg.SetTeardown(l.teardown)
+	l.proc.SetTeardown(l.teardown)
```

Comment:
i think there's a race here. If `ctx` is done _before_ `l.proc.SetTeardown(l.teardown)` is called, we lose the call to the teardown. I think the race was there before, maybe we can fix it as part of this PR.

it may be that we need a special:
```
goprocessctx.WithContextAndTeardown(ctx, l.teardown)
```
function, as you had suggested earlier.